src/QsCompiler/QirGeneration/QIR/DataStructures.cs

```diff
@@ -383,40 +421,46 @@ internal TupleValue(Value tuple, ImmutableArray<ResolvedType> elementTypes, Gene
-                }
-                else
-                {
-                    void Store(IValue v) =>
-                        this.LlvmNativeValue = this.sharedState.CurrentBuilder.InsertValue(this.Value, v.Value, (uint)index);
-
-                    IValue Reload() =>
-                        this.sharedState.Values.From(this.sharedState.CurrentBuilder.ExtractValue(this.Value, (uint)index), type.Item1);
-
-                    return new PointerValue(type.Item1, type.Item2, this.sharedState, Reload, Store);
-                }
-            })));
-
-        private Value AllocateTuple(bool registerWithScopeManager)
+        private PointerValue CreateTupleElementPointer(int index, IValue? element = null)
         {
-            // The runtime function TupleCreate creates a new value with reference count 1 and alias count 0.
-            var constructor = this.sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.TupleCreate);
-            var size = this.sharedState.ComputeSizeForType(this.StructType);
-            var tuple = this.sharedState.CurrentBuilder.Call(constructor, size);
-            if (registerWithScopeManager)
+            if (this.IsRuntimeManaged)
             {
-                this.sharedState.ScopeMgr.RegisterValue(this);
+                var elementPtr = this.sharedState.CurrentBuilder.GetStructElementPointer(this.StructType, this.TypedPointer, (uint)index);
+                return new PointerValue(elementPtr, this.ElementTypes[index], this.LlvmElementTypes[index], this.sharedState);
             }
+            else
+            {
+                void Store(IValue v) =>
+                    this.LlvmNativeValue = this.sharedState.CurrentBuilder.InsertValue(this.Value, v.Value, (uint)index);
+
+                IValue Reload() =>
+                    this.sharedState.Values.From(this.sharedState.CurrentBuilder.ExtractValue(this.Value, (uint)index), this.ElementTypes[index]);
 
-            return tuple;
+                return element is null
+                    ? new PointerValue(this.ElementTypes[index], this.LlvmElementTypes[index], this.sharedState, Reload, Store)
+                    : new PointerValue(element, this.sharedState, Reload, Store);
+            }
         }
 
+        private IValue.Cached<PointerValue> CreateTupleElementPointerCache(int index, PointerValue? pointer = null) =>
+            new(pointer is null || this.IsRuntimeManaged
+                    ? null
+                    : this.CreateTupleElementPointer(index, pointer.CurrentCache()),
+                this.sharedState,
+                () => this.CreateTupleElementPointer(index));
+
+        private IValue.Cached<PointerValue>[] CreateTupleElementPointersCaches() =>
+            Enumerable.ToArray(Enumerable.Range(0, this.ElementTypes.Length).Select(index => this.CreateTupleElementPointerCache(index)));
+
+        private Value AllocateTuple(bool allocOnStack) =>
```

Comment:
`AllocateArray` can only be used for opaque, runtime managed arrays, while `AllocateTuple` takes a boolean to determine if it will stack allocated instead. For arrays, stack allocation appears to be handled by separate functions. While the implementation seems fine, it's a little odd to have this inconsistency. Should tuples also be split out into two different functions rather than use `AllocateTuple` for both?

Reply:
updated so that this function doesn't take an argument